File: workspace/workflows/src/metadata/begin_transition_to.rs
use crate::activity::{Activity, ActivityContext, Error};
use async_trait::async_trait;
use serde_json::Value;
use bosca_client::client::{Client, WorkflowJob};
use bosca_client::client::add_activity::ActivityInput;
use bosca_client::client::begin_transition::BeginTransitionInput;

pub struct MetadataBeginTransitionToActivity {
    id: String,
}

impl Default for MetadataBeginTransitionToActivity {
    fn default() -> Self {
        Self::new()
    }
}

impl MetadataBeginTransitionToActivity {
    pub fn new() -> MetadataBeginTransitionToActivity {
        MetadataBeginTransitionToActivity {
            id: "metadata.begin.transition.to".to_string(),
        }
    }
}

#[async_trait]
impl Activity for MetadataBeginTransitionToActivity {
    fn id(&self) -> &String {
        &self.id
    }

    fn create_activity_input(&self) -> ActivityInput {
        let mut configuration = serde_json::map::Map::new();
        configuration.insert("state".to_string(), Value::String("draft".to_string()));
        configuration.insert("status".to_string(), Value::String("Transitioning to Draft".to_string()));
        ActivityInput {
            id: self.id.to_owned(),
            name: "Begin to Transition Metadata".to_string(),
            description: "Begin to transition the metadata workflow state".to_string(),
            child_workflow_id: None,
            configuration: Value::Object(configuration),
            inputs: vec![],
            outputs: vec![],
        }
    }

    async fn execute(&self, client: &Client, _: &mut ActivityContext, job: &WorkflowJob) -> Result<(), Error> {
        let state = job
            .workflow_activity
            .configuration
            .get("state")
            .unwrap()
            .as_str()
            .unwrap()
            .to_string();
        let status_value = job.workflow_activity.configuration.get("status").unwrap();
        let status = if status_value.is_null() {
            "".to_string()
        } else {
            status_value.as_str().unwrap().to_string()
        };
        let metadata = &job.metadata.clone().unwrap();
        client
            .begin_transition(BeginTransitionInput {
                metadata_id: Some(metadata.id.to_owned()),
                version: Some(metadata.version.to_owned()),
                collection_id: None,
                state_id: state.to_owned(),
                status,
                supplementary_id: None,
                wait_for_completion: None
            })
            .await?;
        Ok(())
    }
}

File: workspace/runners/kotlin/activities/src/main/kotlin/io/bosca/workflow/collection/BeginTransitionTo.kt
package io.bosca.workflow.collection

import io.bosca.api.Client
import io.bosca.graphql.fragment.WorkflowJob
import io.bosca.graphql.type.ActivityInput
import io.bosca.util.toOptional
import io.bosca.workflow.Activity
import io.bosca.workflow.ActivityContext

class BeginTransitionTo(client: Client) : Activity(client) {

    override val id: String = ID

    override suspend fun toActivityDefinition(): ActivityInput {
        return ActivityInput(
            id = id,
            name = "Begin Collection Transition",
            description = "Begin a Collection Transition",
            configuration = mapOf<String, Any>(
                "state" to "draft",
            ).toOptional(),
            inputs = emptyList(),
            outputs = emptyList(),
        )
    }

    override suspend fun execute(context: ActivityContext, job: WorkflowJob) {
        val configuration = job.workflowActivity.workflowActivity.configuration as Map<*, *>
        val state = configuration["state"] as String
        val current = job.collection?.collection?.workflow?.collectionWorkflow
        if (state != current?.state && state != current?.pending) {
            client.workflows.beginCollectionTransition(
                job.collection?.collection?.id ?: error("missing collection"),
                "Begin Collection Transition",
                state
            )
        }
    }

    companion object {
        const val ID = "collection.begin.transition.to"
    }
}